lib/jsonschema-tools.js

```diff
@@ -859,10 +895,30 @@ async function materializeAllSchemas(options = {}) {
 
     return _.flatten(await Promise.all(_.map(
         currentSchemasInfo,
-        info => materializeSchemaVersion(path.dirname(info.path), info.schema, options)
+        info => materializeSchemaToPath(path.dirname(info.path), info.schema, options)
     )));
 }
 
+/**
+ * Traverses through all the fields in the schema, including nested properties, and
+ * sets the numeric bounds specified in options when the object type is number.
+ * @param {Object} schema
+ * @param {Array} bounds as specified in options
+ * @return {Object} schema with bounded numeric fields
+ */
+function enforceNumericBounds(schema, bounds) {
+    const schemaCopy = _.cloneDeep(schema);
+    const enforcedMin = bounds[0];
+    const enforcedMax = bounds[1];
+    traverseSchema(schemaCopy, (obj) => {
+        if (['number', 'integer', 'float'].includes(obj.type)) {
```

Comment:
Is 'float' even a jsonschema type?    I don't think so, just integer and number (which we interpret as a decimal).

https://json-schema.org/understanding-json-schema/reference/numeric.html

Also, don't the numeric bounds need to vary based on whether or not we have an integer or float?  I'd make the default integer be default max,min 64 bit signed Long, and number be default max,min 64 bit signed Double, right?

Reply:
Still need response on this one:

> Also, don't the numeric bounds need to vary based on whether or not we have an integer or float? I'd make the default integer be default max,min 64 bit signed Long, and number be default max,min 64 bit signed Double, right?

Reply:
Sorry for not responding to this @ottomata 

Sooo for floats it's tricky to assign a fixed min and max, because we're not interested anymore in those bounds, but in the precision of the float. The more decimal positions a number has, the narrower is the flat [min, max] numeric bounds (i.e. the earlier the numbers become unsafe). As far as I know in JSONSchema we can only validate numbers through numeric ranges (with the `maximum` and `minimum` props), not through the actual size in memory of the number. So I think we should just keep it simple and let the repo maintainer specify a hard limit on any number, int or float

This is an excellent SO answer that made the issue very clear to my no-computer-science-degree self: [](https://stackoverflow.com/questions/45929493/node-js-maximum-safe-floating-point-number)

Reply:
> So I think we should just keep it simple and let the repo maintainer specify a hard limit on any number, int or float

This could make sense.  It looks like Javascript also has [MAX_VALUE](https://tc39.es/ecma262/#sec-number.max_safe_integer), which lines up with the max integer part of a 64bit double. 

From https://en.wikipedia.org/wiki/Double-precision_floating-point_format:
> The 11 bit width of the exponent allows the representation of numbers between 10^-308 and 10^308, with full 15–17 decimal digits precision. By compromising precision, the subnormal representation allows even smaller values up to about 5 × 10−324.


I had assumed that MAX_SAFE_INTEGER would have been more than the max double value, but it is actually more restrictive.  Since it is more restrictive, I think I agree, let's just use the same bounds for both integers and numeric types, even though those values will eventually be stored in different binary datatypes with different ranges.

TY!